src/Data/Deque/Mutable/Linear.hs

```diff
@@ -0,0 +1,182 @@
+
+-- | Run a computation on a Deque that is deterimined by the given the list
+-- where we treat the start and end of the list as the left and right pointers,
+-- with the total capacity as the length of the list.
+fromList :: [a] -> (Deque a %1-> Ur b) %1-> Ur b
+fromList xs f | l <- Prelude.length xs = Array.fromList xs $
+  \arr -> f (Deque l (Just (0,l-1)) arr)
+
+
+-- # Querying
+-------------------------------------------------------------------------------
+
+-- | The total capacity of the Deque
+size :: Deque a %1-> (Ur Int, Deque a)
+size (Deque sz ptrs arr) = (Ur sz, Deque sz ptrs arr)
+
+-- | The number of elements currently stored
+length :: Deque a %1-> (Ur Int, Deque a)
+length (Deque sz Nothing arr) = (Ur 0, Deque sz Nothing arr)
+length (Deque sz p@(Just (l,r)) arr)
+  | l <= r = (Ur (1 + r - l), Deque sz p arr)
+  | otherwise = (Ur (sz + 1 - l + r), Deque sz p arr)
+
+-- | We are full if the length equals the size
+isFull :: Deque a %1-> (Ur Bool, Deque a)
+isFull d =
+  length d & \(Ur len, Deque sz p arr) -> (Ur (len == sz), Deque sz p arr)
+
+peek :: HasCallStack => Dir -> Deque a %1-> (Ur a, Deque a)
+peek _ (Deque _ Nothing arr) = error "Peeking in empty deque." $ arr
+peek dir (Deque sz p@(Just (l,r)) arr) = case dir of
+  L -> Array.read arr l & \(x, arr') -> (x, Deque sz p arr')
+  R -> Array.read arr r & \(x, arr') -> (x, Deque sz p arr')
+
+-- | View the top of the left queue
+peekLeft :: HasCallStack => Deque a %1-> (Ur a, Deque a)
+peekLeft = peek L
+
+-- | View the top of the right queue
+peekRight :: HasCallStack => Deque a %1-> (Ur a, Deque a)
+peekRight = peek R
+
+
+-- # Modification
+-------------------------------------------------------------------------------
+
+add :: HasCallStack => Dir -> Deque a %1-> a -> Deque a
+add dir deq x = isFull deq & \case
+  (Ur True, deq') -> error "Adding to a full deque" $ deq'
+  (Ur False, Deque sz Nothing arr) ->
+    Deque sz (Just (0,0)) (Array.write arr 0 x)
+  (Ur False, Deque sz (Just (l,r)) arr) -> case dir of
+    L -> Array.write arr (prevIx sz l) x &
+      \arr' -> Deque sz (Just (prevIx sz l, r)) arr'
+    R -> Array.write arr (nextIx sz r) x &
+      \arr' -> Deque sz (Just (l,nextIx sz r)) arr'
+
+-- | Add to the left queue
+addLeft :: HasCallStack => Deque a %1-> a -> Deque a
+addLeft = add L
+
+-- | Add to the right queue
+addRight :: HasCallStack => Deque a %1-> a -> Deque a
+addRight = add R
+
+pop :: HasCallStack => Dir -> Deque a %1-> Deque a
+pop _ (Deque _ Nothing arr) = error "Popping from empty deque." $ arr
+pop dir (Deque sz (Just (l,r)) arr)
+  | l==r = Deque sz Nothing arr
+  | otherwise = case dir of
+    L -> Deque sz (Just (nextIx sz l, r)) arr
+    R -> Deque sz (Just (l, prevIx sz r)) arr
+
+-- | Remove the last added element from the left queue
+popLeft :: HasCallStack => Deque a %1-> Deque a
+popLeft = pop L
+
+-- | Remove the last added element from the right queue
+popRight :: HasCallStack => Deque a %1-> Deque a
+popRight = pop R
```

Comment:
Instead of `Left`/`Right`, I would refer to them as `Front`/`Back` as it is more common.

Reply:
Good idea :+1: